### PR TITLE
Experimental route util API: check if route has unexpected closures

### PR DESCRIPTION
**Pull request by RingerJK**
### Description

Experimental route util API: check if route has unexpected closures

~TODO: need docs~

### CODE REVIEW

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * The method checks whenever the route has unexpected closures based
+     * on [RouteOptions.snappingIncludeClosuresList] and [RouteOptions.snappingIncludeStaticClosuresList]
+     * flags. If the any of them is **false** for the waypoint and closure overlaps it the method
+     * immediately returns **true**.
+     *
+     * If route has closures that are not overlapped waypoints, the method immediately returns
+     * **true**.
+     *
+     * Note: Experimental preview API
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:31]:
```suggestion
     * This function checks whether the [navigationRoute] has unexpected closures, which could be a reason to re-route.
     *
     * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
     * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
     * flags. Otherwise, `false` is returned.
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    @ExperimentalPreviewMapboxNavigationAPI
+    fun hasRouteUnexpectedClosures(navigationRoute: NavigationRoute): Boolean {
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:34]:
```suggestion
    fun NavigationRoute.hasUnexpectedClosures(): Boolean {
```
and move to a public `NavigationRouteEx` file (with JVM file name `NavigationRouteUtils`).

It will also need to be a suspend/callback function (or be instructed to be used from a worker thread) because parsing geometries can be expensive for long routes.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        val points = mutableSetOf<Point>()
+
+        steps()?.forEach { step ->
+            step.geometry()?.let { geometry ->
+                points.addAll(PolylineUtils.decode(geometry, precision))
+            } ?: run {
+                logE("LegStep doesn't have geometry", LOG_CATEGORY)
+            }
+        }
+        return points.toList()
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:150]: This doesn't deduplicate points, so the duplicates at start and ends of steps will inflate the indices. Also, the `DecodeUtils` should be used (on a worker) so that the cache can be reused across the app.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                )
+
+                if (closure.geometryIndexStart() == 0 &&
+                    snappingResultList.getOrNull(index) != true
+                ) {
+                    logW(
+                        "Route with id [${navigationRoute.id}] has closure at the start of the" +
+                            " leg, leg index $routeLegIndex",
+                        LOG_CATEGORY
+                    )
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:59]: No need for a log, this is an expected result of the function. Same below.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/util/route/RouteUtilTest.kt [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                return mockk {
+                    every { directionsRoute } returns route
+                    every { id } returns "-1"
+                }
+            }
+
+            private fun provideBooleansProvider(
+                vararg bool: Boolean?,
+            ): BooleansProvider = { bool.asList() }
+        }
```
Comment (VysotskiVadim) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/util/route/RouteUtilTest.kt:145]: maybe that's simpler? 
```suggestion
            private fun mockNavigationRoute(
                directionResponse: String,
                snappingIncludeClosures: List<Boolean>?,
                snappingIncludeStaticClosures: List<Boolean>?,
            ): NavigationRoute {
                return createNavigationRoutes(
                    DirectionResponse.fromJson(directionResponse),
                    createRouteOptions(
                        snappingIncludeClosuresList = snappingIncludeClosures,
                        snappingIncludeStaticClosures = snappingIncludeStaticClosures,
                        TestSDKRouteParser(),
                        com.mapbox.navigation.base.route.RouterOrigin.Offboard
                ).first()
        }
```

P.S. I understand that given the time pressure you may don't have time to even discuss it, feel free to merge and get back to this question later 🙂

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    private fun DirectionsRoute.getSnappingResultList(): List<Boolean> {
+        val snappingIncludeClosuresList = routeOptions()?.snappingIncludeClosuresList()
+        val snappingIncludeStaticClosuresList = routeOptions()?.snappingIncludeStaticClosuresList()
+
+        val snappingResultList = mutableListOf<Boolean>()
+        for (
+            index in 0 until maxOf(
+                snappingIncludeClosuresList?.size ?: 0,
+                snappingIncludeStaticClosuresList?.size ?: 0
```
Comment (VysotskiVadim) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:111]: just curious, is it possible for them to have a different size?

Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:111]: No, a request would fail.

However, one can be `null` and the other filled up.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt (hunk 4 above)
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt:59]: let's keep it with `logD` to have more info in case of any issues

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/util/route/RouteUtilTest.kt (hunk 5 above)
Comment (RingerJK) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/util/route/RouteUtilTest.kt:145]: yeah, it also would work 👍 let's keep the same, as we want to include this to release today

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+ * This function checks whether the [NavigationRoute] has unexpected closures, which could be a reason to re-route.
+ *
+ * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
+ * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
+ * flags. Otherwise, `false` is returned.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+suspend fun NavigationRoute.hasUnexpectedClosures(): Boolean =
+    withContext(Dispatchers.IO) {
+        val LOG_CATEGORY = "HAS_ROUTE_UNEXPECTED_CLOSURES"
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:26]:
```suggestion
        val LOG_CATEGORY = "NavigationRouteUtils"
```
and move to a `private const val`

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    return mapIndexedNotNull { index, legSilentWaypoints ->
+        if (legSilentWaypoints.geometryIndex in geometryIndexStart..geometryIndexEnd) {
+            index
+        } else {
+            null
+        }
+    }
+}
+
+private fun DirectionsRoute.stepsGeometryToPoints(
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:129]:
```suggestion
private fun stepsGeometryToPoints(
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+/**
+ * This function checks whether the [NavigationRoute] has unexpected closures, which could be a reason to re-route.
+ *
+ * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
+ * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
+ * flags. Otherwise, `false` is returned.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+suspend fun NavigationRoute.hasUnexpectedClosures(): Boolean =
+    withContext(Dispatchers.IO) {
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:25]:
```suggestion
    withContext(Dispatchers.Default) {
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    leg.steps()?.forEachIndexed { stepIndex, legStep ->
+        stepGeometryToPoints(legStep).let {
+            // step with 2 coordinate might have duplicated values
+            val squashed = if (it.size == 2) {
+                it.toSet()
+            } else {
+                it
+            }
+            if (stepIndex != 0) {
+                squashed.drop(1)
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:143]:
```suggestion
                // removing duplicate points for adjacent steps
                squashed.drop(1)
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at leg " +
+                            "index $routeLegIndex",
+                    )
+                    return@withContext true
+                }
+            }
+            index += silentWaypoints.size
+        }
+
+        return@withContext false
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:91]: This should be flipped.

```suggestion
        return@withContext true
```

The goal is to do as little checks as possible to confirm that a closure is allowed, and only when we don't find any acceptable criteria return `true`.

This allows us to, for example, first check for the silent waypoints (which is the cheapest check because it doesn't require deserialization of geometries), and only check for all regular waypoints, and ultimately fail if no criteria was found.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt (hunk 8 above)
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:129]: `fun stepGeometryToPoints` requires `DirectionsRoute`

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt (hunk 11 above)
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:91]: must be returned `false`: it means that the route `doesn't` have unexpected closures.

>This allows us to, for example, first check for the silent waypoints (which is the cheapest check because it doesn't require deserialization of geometries), and only check for all regular waypoints, and ultimately fail if no criteria was found.

we have to measure geometry before the `for loop`, so geometry still is measured, event if we check silent waypoints first.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                        provideBooleansProvider(false, true, true),
+                        true,
+                    ),
+                    arrayOf(
+                        "route closure at starting waypoint, include closures = true",
+                        "route_closure_start_coordinate.json",
+                        provideBooleansProvider(true, false, false),
+                        provideBooleansProvider(true, false, false),
+                        false,
+                    ),
```
Comment (LukasPaczos) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:55]:
```suggestion
                    arrayOf(
                        "route closure at starting waypoint, include closures = true",
                        "route_closure_start_coordinate.json",
                        provideBooleansProvider(true, false, false),
                        provideBooleansProvider(true, false, false),
                        false,
                    ),
                    arrayOf(
                        "route closure at starting waypoint, include closures = true",
                        "route_closure_start_coordinate.json",
                        provideBooleansProvider(false, false, false),
                        provideBooleansProvider(true, false, false),
                        false,
                    ),
                    arrayOf(
                        "route closure at starting waypoint, include closures = true",
                        "route_closure_start_coordinate.json",
                        provideBooleansProvider(true, false, false),
                        provideBooleansProvider(false, false, false),
                        false,
                    ),
```
This is also a wariant we need to check, don't we?

File: CHANGELOG.md [hunk 13]
```diff
@@ -20,6 +20,7 @@ Mapbox welcomes participation and contributions from everyone.
 - Fixed an issue with `NavigationView` that caused camera to unexpectedly change state in some situations. [#6291](https://github.com/mapbox/mapbox-navigation-android/pull/6291)
 - Fixed an issue where step distance in `MapboxManeuverView` could have been cut off. [#6296](https://github.com/mapbox/mapbox-navigation-android/pull/6296)
 - Fixed an issue where trip progress in `MapboxTripProgressView` could have been cut off if large font size is used. [#6296](https://github.com/mapbox/mapbox-navigation-android/pull/6296)
+- Added Experimental API `NavigationRoute#hasUnexpectedClosures` to check if a route has unexpected closures and it's required to be re-routed. [#6295](https://github.com/mapbox/mapbox-navigation-android/pull/6295)
```
Comment (LukasPaczos) [inline CHANGELOG.md:23]:
```suggestion
- Added Experimental API `NavigationRoute#hasUnexpectedClosures` to check if a route has unexpected closures that might require a re-route. [#6295](https://github.com/mapbox/mapbox-navigation-android/pull/6295)
```

**Review by LukasPaczos — APPROVED** [review]
Two minor comments above. Otherwise, let's ship for first iteration and we can look into other potential edge cases.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt (hunk 12 above)
Comment (RingerJK) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:55]: I can add something like this, but not for every route to avoid overkilling

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * return the list of indexes of List<LegSilentWaypoints> which are at closure area
+ */
+private fun List<LegSilentWaypoints>.inGeometryRange(
+    geometryIndexStart: Int,
+    geometryIndexEnd: Int
+): List<Int> {
+    return mapIndexedNotNull { index, legSilentWaypoints ->
+        if (legSilentWaypoints.geometryIndex in geometryIndexStart..geometryIndexEnd) {
```
Comment (dzinad) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:127]: Is it supposed to be inclusive? It's hard to tell from `Closure` docs but looks like so, just clarifying.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+                directionsRoute.stepsGeometryToPoints(routeLeg).lastIndex
+            }
+
+            val silentWaypoints = routeLeg.silentWaypoints()
+
+            routeLeg.closures()?.forEach { closure ->
+                val silentWaypointsInClosureRange = silentWaypoints.inGeometryRange(
+                    closure.geometryIndexStart(), closure.geometryIndexEnd()
+                )
+                if (silentWaypointsInClosureRange.isNotEmpty()) {
```
Comment (dzinad) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:43]: Why do you need this check? Can't you just rely on forEach not being executed for an empty collection?

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+                    arrayOf(
+                        "route closure at starting waypoint, include closures = true",
+                        "route_closure_start_coordinate.json",
+                        provideBooleansProvider(false, false, false),
+                        provideBooleansProvider(true, false, false),
+                        false,
+                    ),
+                    arrayOf(
+                        "route closure at starting waypoint, include closures = true",
```
Comment (dzinad) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:64]: This description is the same as the two descriptions above.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+                    arrayOf(
+                        "route closure at second waypoint, include closures = true",
+                        "route_closure_second_waypoint.json",
+                        provideBooleansProvider(false, true, false),
+                        provideBooleansProvider(false, true, false),
+                        false,
+                    ),
+                    arrayOf(
+                        "route closure at second *silent* waypoint, include closures = false",
+                        "route_closure_second_silent_waypoin.json",
```
Comment (dzinad) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:86]: Typo.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+                        provideBooleansProvider(true, false, true),
+                        provideBooleansProvider(true, false, true),
+                        true,
+                    ),
+                    arrayOf(
+                        "route closure at second *silent* waypoint, include closures = true",
+                        "route_closure_second_silent_waypoin.json",
```
Comment (dzinad) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:93]: Typo.

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+        private val routeRaw: String,
+        private val snappingIncludeClosures: BooleansProvider?,
+        private val snappingIncludeStaticClosures: BooleansProvider?,
+        private val expectedHasUnexpectedClosures: Boolean,
+    ) {
+
+        companion object {
+            @JvmStatic
+            @Parameterized.Parameters(name = "{0}")
+            fun data(): List<Array<Any?>> =
```
Comment (dzinad) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:33]: Do you test anywhere null booleans in closures list? I see only true/false.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * This function checks whether the [NavigationRoute] has unexpected closures, which could be a reason to re-route.
+ *
+ * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
+ * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
+ * flags. Otherwise, `false` is returned.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+suspend fun NavigationRoute.hasUnexpectedClosures(): Boolean =
```
[the remark this answers is not in the document]
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:26]: For the experimental API, we should be good to have suspended only.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+private fun DirectionsRoute.stepsGeometryToPoints(
+    leg: RouteLeg,
+): List<Point> {
+    val points = mutableListOf<Point>()
+
+    leg.steps()?.forEachIndexed { stepIndex, legStep ->
+        stepGeometryToPoints(legStep).let {
+            // step with 2 coordinate might have duplicated values
+            val squashed = if (it.size == 2) {
```
[the remark this answers is not in the document]
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:143]: I don't think so, but we can clarify. @purew could you provide, is it possible `>2` **in a row** duplicated points in a step?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+    return mapIndexedNotNull { index, legSilentWaypoints ->
+        if (legSilentWaypoints.geometryIndex in geometryIndexStart..geometryIndexEnd) {
+            index
+        } else {
+            null
+        }
+    }
+}
+
+private fun DirectionsRoute.stepsGeometryToPoints(
```
[the remark this answers is not in the document]
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:135]: Actually, I use it, see the line `stepGeometryToPoints(legStep).let {`. I needed to clean up duplicated `Point`s. It would be nice to move this to extensions of `DecodeUtils`. Let's address it separately when we get that it works good

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt (hunk 15 above)
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt:43]: yes, it can. Non needed check 👍

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt (hunk 19 above)
Comment (RingerJK) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:33]: yeah, that make sense. Cut the issue https://github.com/mapbox/mapbox-navigation-android/issues/6312

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt (hunk 17 above)
Comment (RingerJK) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:86]: https://github.com/mapbox/mapbox-navigation-android/issues/6312

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt (hunk 16 above)
Comment (RingerJK) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/utils/route/NavigationRouteExTest.kt:64]: there's only changed OTHER closures boolean. Should be address with https://github.com/mapbox/mapbox-navigation-android/pull/6295#discussion_r968239294
